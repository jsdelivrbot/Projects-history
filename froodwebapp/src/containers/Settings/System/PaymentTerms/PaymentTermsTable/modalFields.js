export default fromFields => (
  [{
    header: 'Term',
    type: 'text',
    name: 'name',
    editableForExistingItem: false,
    editableForNewItem: true
  }, {
    header: 'Due In Days',
    type: 'text',
    name: 'dueDays',
    editableForExistingItem: true,
    editableForNewItem: true
  }, {
    header: 'From',
    type: 'select',
    name: 'from',
    editableForExistingItem: true,
    editableForNewItem: true,
    menuItems: fromFields.map(field => ({ key: field.id, value: field.name }))
  }]
);
