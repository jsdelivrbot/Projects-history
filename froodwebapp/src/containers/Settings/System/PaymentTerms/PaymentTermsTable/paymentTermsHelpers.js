import React from 'react';
import { TableActionButtons } from 'components';

export const columns = (handleEdit, handleDeactivate, handleActivate) => [{
  title: 'Terms',
  dataIndex: 'name',
  width: '33%'
}, {
  title: 'Due In Days',
  dataIndex: 'dueDays',
  width: '33%'
}, {
  title: 'Action',
  width: '33%',
  render: (text, record) => (
    <TableActionButtons
      record={ record }
      handleEdit={ handleEdit }
      handleDeactivate={ handleDeactivate }
      handleActivate={ handleActivate }
      activateButtonVisible
    />
  )
}];

// add string to dueDays
export const addStringsToDueDays = (paymentTerms, fromFields) => (
  paymentTerms.map(pt => ({
    id: pt.id,
    name: pt.name,
    isActive: pt.isActive,
    dueDays: `${pt.dueDays} Days from ${fromFields.find(field => field.id === Number(pt.from)).name}`,
  }))
);
