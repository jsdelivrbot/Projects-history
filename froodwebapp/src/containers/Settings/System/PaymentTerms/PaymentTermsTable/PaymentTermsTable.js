import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Row, Col } from 'react-flexbox-grid';
import { connect } from 'react-redux';
import {
  payTermsGetRequest,
  payTermsSaveRequest,
  payTermsUpdateRequest,
  payTermsDeleteRequest
} from 'redux-base/actions';
import { Table, Spin } from 'antd';
import { ConnectedTopFormModal, FroodSearchInput } from 'components';
import { table } from 'styles/common.scss';
import { columns, addStringsToDueDays } from './paymentTermsHelpers';
import fields from './modalFields';

const mapStateToProps = state => ({
  needReloadPayTerms: state.paymentTerms.needReloadPayTerms,
  loadingPage: state.paymentTerms.loadingPage,
  paymentTerms: state.paymentTerms.data,
  fromFields: state.paymentTerms.from,
});

const mapDispatchToProps = {
  payTermsGetRequest,
  payTermsSaveRequest,
  payTermsUpdateRequest,
  payTermsDeleteRequest
};

const newTerm = {
  name: '',
  dueDays: '',
  from: ''
};

export class PaymentTermsTable extends Component {
  constructor(props) {
    super(props);
    this.state = {
      modalVisible: false,
      modalData: newTerm,
      searchValue: ''
    };
  }

  componentWillMount() {
    this.props.payTermsGetRequest();
  }

  componentWillReceiveProps(nextProps) {
    // when we delete or update term
    if (nextProps.needReloadPayTerms) {
      this.setState({
        modalVisible: false,
        modalData: undefined
      }, () => this.props.payTermsGetRequest());
      return;
    }

    // when we create term
    if (this.props.paymentTerms.length !== nextProps.paymentTerms.length) {
      this.setState({
        modalVisible: false,
        modalData: newTerm
      });
    }

    // add string to dueDays
    nextProps.paymentTerms.filter((item) => {
      const data = item;
      const { name } = nextProps.fromFields.find(field => field.id === Number(item.from));
      data.dueDaysString = `${data.dueDays} Days from ${name}`;
      return data;
    });
  }

  handleToggleModal = () => {
    this.setState({
      modalVisible: !this.state.modalVisible,
      modalData: newTerm
    });
  }

  handleSave = (term) => {
    if (this.state.modalData.id) {
      this.props.payTermsUpdateRequest({
        payload: {
          id: this.state.modalData.id,
          dueDays: term.dueDays,
          from: term.from
        }
      });
    } else {
      this.props.payTermsSaveRequest({
        payload: {
          name: term.name,
          dueDays: Number(term.dueDays),
          from: Number(term.from),
        }
      });
    }
  }

  handleActivate = (e) => {
    const termId = e.target.id;
    const { dueDays } = this.props.paymentTerms.find(term => term.id === Number(termId));
    this.props.payTermsUpdateRequest({
      payload: {
        id: termId,
        dueDays
      }
    });
  }

  handleDeactivate = (e) => {
    this.props.payTermsDeleteRequest({
      id: e.target.id
    });
  }

  handleEdit = (e) => {
    this.setState({
      modalVisible: true,
      modalData: this.props.paymentTerms.find(pt => pt.id === Number(e.target.id))
    });
  }

  handleSearchChange = (value) => {
    this.setState({
      searchValue: value
    });
  }

  render() {
    const {
      modalData,
      modalVisible,
      searchValue
    } = this.state;

    const {
      paymentTerms,
      fromFields,
      loadingPage
    } = this.props;

    const dataTerms = addStringsToDueDays(paymentTerms, fromFields);

    const data = dataTerms.filter(item =>
      item && item.name && item.name.toLowerCase().includes(searchValue.toLowerCase())
    );

    return (
      <Row>
        <Col lg>
          <ConnectedTopFormModal
            loading={ loadingPage }
            title="New Term"
            buttonText="New Term"
            handleSave={ this.handleSave }
            handleToggleModal={ this.handleToggleModal }
            visible={ modalVisible }
            initialValues={ modalData }
            fields={ fields(fromFields) }
          />
          <Spin spinning={ loadingPage }>
            <Row>
              <Col lg={ 4 }>
                <FroodSearchInput
                  onChange={ this.handleSearchChange }
                  placeholder="Type to search payment terms"
                />
              </Col>
            </Row>
            <Table
              className={ table }
              rowKey="id"
              size="small"
              columns={
                columns(
                  this.handleEdit,
                  this.handleDeactivate,
                  this.handleActivate) }
              dataSource={ data }
            />
          </Spin>
        </Col>
      </Row>
    );
  }
}

PaymentTermsTable.propTypes = {
  // trigger
  loadingPage: PropTypes.bool.isRequired,
  needReloadPayTerms: PropTypes.bool.isRequired,
  // data
  paymentTerms: PropTypes.array.isRequired,
  fromFields: PropTypes.array.isRequired,
  // redux-base
  payTermsGetRequest: PropTypes.func.isRequired,
  payTermsSaveRequest: PropTypes.func.isRequired,
  payTermsUpdateRequest: PropTypes.func.isRequired,
  payTermsDeleteRequest: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(PaymentTermsTable);
