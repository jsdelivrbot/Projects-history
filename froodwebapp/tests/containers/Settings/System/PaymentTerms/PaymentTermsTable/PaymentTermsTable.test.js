import { PaymentTermsTable } from 'containers/Settings/System/PaymentTerms/PaymentTermsTable/PaymentTermsTable';
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const { describe, expect, it } = global;

describe('PaymentTermsTable', () => {
  const props = {
    loadingPage: false,
    needReloadPayTerms: false,
    paymentTerms: [],
    fromFields: [],
    payTermsGetRequest: jest.fn(),
    payTermsSaveRequest: jest.fn(),
    payTermsUpdateRequest: jest.fn(),
    payTermsDeleteRequest: jest.fn()
  };

  it('renders the PaymentTermsTable component', () => {
    const component = shallow(<PaymentTermsTable { ...props } />);

    expect(component.length).toEqual(1);
  });

  it('capturing Snapshot of PaymentTermsTable', () => {
    const component = shallow(<PaymentTermsTable { ...props } />);
    const view = toJson(component);
    const actual = expect(view);

    actual.toMatchSnapshot();
  });
});
