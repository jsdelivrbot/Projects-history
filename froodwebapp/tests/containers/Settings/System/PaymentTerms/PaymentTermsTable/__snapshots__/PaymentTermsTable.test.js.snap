// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentTermsTable capturing Snapshot of PaymentTermsTable 1`] = `
<Row>
  <Col
    lg={true}
  >
    <ReduxForm
      buttonText="New Term"
      fields={
        Array [
          Object {
            "editableForExistingItem": false,
            "header": "Term",
            "name": "name",
            "type": "text",
          },
          Object {
            "editableForExistingItem": true,
            "header": "Due In Days",
            "name": "dueDays",
            "type": "text",
          },
          Object {
            "editableForExistingItem": true,
            "header": "From",
            "menuItems": Array [],
            "name": "from",
            "type": "select",
          },
        ]
      }
      handleSave={[Function]}
      handleToggleModal={[Function]}
      initialValues={
        Object {
          "dueDays": "",
          "from": "",
          "name": "",
        }
      }
      loading={false}
      title="New Term"
      visible={false}
    />
    <Spin
      prefixCls="ant-spin"
      size="default"
      spinning={false}
      wrapperClassName=""
    >
      <Row>
        <Col
          lg={4}
        >
          <FroodSearchInput
            onChange={[Function]}
            placeholder="Type to search payment terms"
          />
        </Col>
      </Row>
      <Table
        bordered={false}
        className="table"
        columns={
          Array [
            Object {
              "dataIndex": "name",
              "title": "Terms",
              "width": "33%",
            },
            Object {
              "dataIndex": "dueDays",
              "title": "Due In Days",
              "width": "33%",
            },
            Object {
              "render": [Function],
              "title": "Action",
              "width": "33%",
            },
          ]
        }
        dataSource={Array []}
        indentSize={20}
        loading={false}
        locale={Object {}}
        prefixCls="ant-table"
        rowKey="id"
        rowSelection={null}
        showHeader={true}
        size="small"
        useFixedHeader={false}
      />
    </Spin>
  </Col>
</Row>
`;
